plugins/out_stackdriver/stackdriver_operation.c

```diff
@@ -64,81 +64,100 @@ int extract_operation(flb_sds_t *operation_id, flb_sds_t *operation_producer,
+    for (; p < pend && op_status == NO_OPERATION; ++p) {
+        key_name = flb_sds_create_len(p->key.via.str.ptr, p->key.via.str.size);
 
-                /* Validate the subfields of operation */
-                for (; tmp_p < tmp_pend; ++tmp_p) {
-                    if (tmp_p->key.type != MSGPACK_OBJECT_STR) {
-                        continue;
-                    }
-                    if (strncmp("id", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_STR) {
-                            continue;
-                        }
-                        *operation_id = flb_sds_copy(*operation_id, tmp_p->val.via.str.ptr, tmp_p->val.via.str.size);
-                    }
-                    else if (strncmp("producer", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_STR) {
-                            continue;
-                        }
-                        *operation_producer = flb_sds_copy(*operation_producer, tmp_p->val.via.str.ptr, tmp_p->val.via.str.size);
-                    }
-                    else if (strncmp("first", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_BOOLEAN) {
-                            continue;
-                        }
-                        if (tmp_p->val.via.boolean) {
-                            *operation_first = FLB_TRUE;
-                        }
-                    }
-                    else if (strncmp("last", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_BOOLEAN) {
-                            continue;
-                        }
-                        if (tmp_p->val.via.boolean) {
-                            *operation_last = FLB_TRUE;
-                        }
-                    }
-                    else {
-                        /* extra sub-fields */ 
-                        *extra_subfields += 1;
-                    }
+        if (p->val.type != MSGPACK_OBJECT_MAP || p->key.type != MSGPACK_OBJECT_STR
+            || strcmp(OPERATION_FIELD_IN_JSON, key_name) != 0) {
+            flb_sds_destroy(key_name);
+            continue;
+        }
+        flb_sds_destroy(key_name);
 
+        op_status = OPERATION_EXISTED;
+        msgpack_object sub_field = p->val;
+        
+        msgpack_object_kv *tmp_p = sub_field.via.map.ptr;
+        msgpack_object_kv *const tmp_pend = sub_field.via.map.ptr + sub_field.via.map.size;
+
+        /* Validate the subfields of operation */
+        for (; tmp_p < tmp_pend; ++tmp_p) {
+            if (tmp_p->key.type != MSGPACK_OBJECT_STR) {
+                continue;
+            }
+            sub_key_name = flb_sds_create_len(tmp_p->key.via.str.ptr, tmp_p->key.via.str.size);
+            if (strcmp("id", sub_key_name) == 0) {
```

Comment:
use strncmp and avoid the memory allocation above

Reply:
done